### PR TITLE
Improve the check_pat_enum logic

### LINKED ISSUE #32004: Can omit fields in pattern match
This code compiles on nightly:

```rust
enum Foo {
    Bar(i32),
    Baz
}

fn main() {
    let x = Foo::Bar(1);
    
    if let Foo::Bar = x {
        println!("Hello");
    }
}
```

Which it shouldn't, since `Foo::Bar` variant has a field and the pattern doesn't mention it. This can also cause an ICE:

```
enum Foo {
    Bar(i32),
    Baz
}

fn main() {
    match Foo::Bar(42) {
        Foo::Bar(x) => {},
        Foo::Bar => {},
        _ => {}
    }
}
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #32004 

r? @eddyb

### CODE REVIEW

File: src/test/compile-fail/issue-32004.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    Bar(i32),
+    Baz
+}
+
+struct S;
+
+fn main() {
+    match Foo::Baz {
+        Foo::Bar => {}
+        //~^ ERROR this pattern has 0 fields, but the corresponding
```
Comment (nagisa) [inline src/test/compile-fail/issue-32004.rs]: Could you expand the expected message to also include “variant” here?

File: src/test/compile-fail/issue-32004.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        _ => {}
+    }
+
+    match S {
+        S(()) => {}
+        //~^ ERROR this pattern has 1 field, but the corresponding
```
Comment (nagisa) [inline src/test/compile-fail/issue-32004.rs]: And “struct” here?

Comment (arielb1) [inline src/test/compile-fail/issue-32004.rs]: fixed

**Comment by nagisa** [comment]
@bors r+ f40d12f

Looks good to me.

**Comment by nagisa** [comment]
I nominate this for beta since this fixes a regression from stable to nightly (now beta) and we recently had a release.

**Pull request merged**